core/arch/arm/kernel/secure_partition.c

```diff
@@ -0,0 +1,751 @@
+		.sp_shared_buf_size = stmm_sec_buf_size,
+		.num_sp_mem_regions = 6,
+		.num_cpus = 1,
+		.mp_info = mp_info,
+	};
+	mp_info->mpidr = read_mpidr_el1();
+	mp_info->linear_id = 0;
+	mp_info->flags = MP_INFO_FLAG_PRIMARY_CPU;
+	spc->ns_comm_buf_addr = comm_buf_addr;
+	spc->ns_comm_buf_size = stmm_ns_comm_buf_size;
+
+	init_stmm_regs(spc, sec_buf_addr,
+		       (vaddr_t)(mp_info + 1) - sec_buf_addr,
+		       stack_addr + stmm_stack_size, image_addr);
+
+	return sec_part_enter_user_mode(spc);
+}
+
+TEE_Result sec_part_init_session(const TEE_UUID *uuid,
+				 struct tee_ta_session *sess)
+{
+	struct sec_part_ctx *spc = NULL;
+	TEE_Result res = TEE_SUCCESS;
+
+	if (memcmp(uuid, &stmm_uuid, sizeof(*uuid)))
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	spc = sec_part_alloc_ctx(uuid);
+	if (!spc)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	spc->is_initializing = true;
+
+	sess->ctx = &spc->uctx.ctx;
+	tee_ta_push_current_session(sess);
+	res = load_stmm(spc);
+	tee_ta_pop_current_session();
+	tee_mmu_set_ctx(NULL);
+	if (res)
+		goto err;
+
+	spc->is_initializing = false;
+	TAILQ_INSERT_TAIL(&tee_ctxes, &spc->uctx.ctx, link);
+	return TEE_SUCCESS;
+
+err:
+	sess->ctx = NULL;
+	spc->uctx.ctx.ops->destroy(&spc->uctx.ctx);
+
+	return res;
+}
+
+static TEE_Result stmm_enter_open_session(struct tee_ta_session *s,
+					  struct tee_ta_param *param __unused,
+					  TEE_ErrorOrigin *eo)
+{
+	struct sec_part_ctx *spc = to_sec_part_ctx(s->ctx);
+
+	if (spc->is_initializing) {
+		/* stmm is initialized in sec_part_init_session() */
+		*eo = TEE_ORIGIN_TEE;
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stmm_enter_invoke_cmd(struct tee_ta_session *s,
+					uint32_t cmd,
+					struct tee_ta_param *param,
+					TEE_ErrorOrigin *eo __unused)
+{
+	struct sec_part_ctx *spc = to_sec_part_ctx(s->ctx);
+	TEE_Result res = TEE_SUCCESS;
+	TEE_Result tmp_res = TEE_SUCCESS;
+	unsigned int ns_buf_size = 0;
+	struct param_mem *mem = NULL;
+	void *va = NULL;
+
+	if (cmd != PTA_STMM_COMMUNICATE)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mem = &param->u[0].mem;
+	ns_buf_size = mem->size;
+	if (ns_buf_size > spc->ns_comm_buf_size)
+		return TEE_ERROR_SHORT_BUFFER;
+
+	res = mobj_inc_map(mem->mobj);
+	if (res)
+		return res;
+
+	va = mobj_get_va(mem->mobj, mem->offs);
+	if (!va) {
+		EMSG("Can't get a valid VA for NS buffer\n");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	spc->regs.x[0] = FFA_MSG_SEND_DIRECT_REQ_64;
+	spc->regs.x[1] = (stmm_pta_id << 16) | stmm_id;
+	spc->regs.x[2] = FFA_PARAM_MBZ;
+	spc->regs.x[3] = spc->ns_comm_buf_addr;
+	spc->regs.x[4] = ns_buf_size;
+	spc->regs.x[5] = 0;
+	spc->regs.x[6] = 0;
+	spc->regs.x[7] = 0;
+
+	tee_ta_push_current_session(s);
+
+	memcpy((void *)spc->ns_comm_buf_addr, va, ns_buf_size);
+
+	res = sec_part_enter_user_mode(spc);
+	/*
+	 * Copy the SPM response from secure partiton back to the non secure
+	 * the client that called us
+	 */
+	param->u[1].val.a = spc->regs.x[4];
+	if (res)
```

Comment:
This should be done before the line above as en error here may mean undefined value in `spc->regs.x[4]`.